lapce-data/src/document.rs

```diff
@@ -1364,20 +1365,29 @@ impl Document {
         self.apply_deltas(&deltas)
     }
 
-    pub fn styles(&self) -> Option<&Arc<Spans<Style>>> {
-        let styles = self
-            .semantic_styles
-            .as_ref()
-            .or_else(|| self.syntax().and_then(|s| s.styles.as_ref()));
-        styles
+    pub fn styles(&self) -> Option<Arc<Spans<Style>>> {
+        let semantic_styles = self.semantic_styles.as_ref();
+        let syntactic_styles = self.syntax().and_then(|s| s.styles.as_ref());
+
+        match (semantic_styles, syntactic_styles) {
+            (Some(semantic), Some(syntax)) => {
+                Some(Arc::new(syntax.merge(semantic, |a, b| match b {
```

Comment:
I think this merge needs to be happening when semantic styles and tree sitter styles get updated. 

And we can just replace semantic styles with the merged styles.

Reply:
Are those changes alright for now?

If we notice perf issues, we can use https://github.com/xi-editor/xi-editor/blob/master/rust/rope/src/spans.rs#L193-L287


As a more efficient solution (that way we only build the spans once instead of 3x. The reason it would be a pain to do now is that the spans have to be added sequentially which isn't the easiest to do